In [1]:
import os

In [10]:
with open("sherlock.txt", "r") as file:
    text = file.read().split('\n')

In [11]:
print(len(text))

76765


We see that there's only 76765 lines in this text. Because the structure of the text is too unstructrued to parse using a single logic, we'll do the dirty job to parse them work by work

In [132]:
records = list([])

In [128]:
# A study in scarlet
scarlet = text[98:4956]
with open('A Study in Scarlet'.upper() + '.txt', "w") as file:
    file.write("\n".join(scarlet))

In [123]:
def get_chapters(story, part_name, part, no_chapter=False, chapter_name=None):
    chapter_list = list([])
    if no_chapter:
        chapter_list.append(dict({"story": story,
                                  "part": part_name,
                                  "chapter": "NO CHAPTER" if chapter_name is None else chapter_name,
                                  "text": list([])}))
        for i in range(len(part)):
            line = part[i].strip()
            if len(line) > 0 and (i==0 or part[i-1].strip() == ""):
                chapter_list[-1]["text"].append(line)
            elif len(line) > 0 and part[i-1] != "":
                chapter_list[-1]["text"][-1] += (" " + line) 
    for i in range(len(part)):
        line = part[i].strip()
        if "CHAPTER" in line:
            chapter_name = " ".join([line, part[i+1].strip()]) if part[i+1].strip()!="" else line
            print(chapter_name)
            chapter_list.append(dict({"story": story,
                                      "part": part_name,
                                      "chapter": chapter_name, 
                                      "text": list([])}))
        elif i==0 or "CHAPTER" in part[i-1]:
            continue
        elif len(line) > 0 and part[i-1] == "":
            chapter_list[-1]["text"].append(line)
        elif len(line) > 0 and part[i-1] != "":
            chapter_list[-1]["text"][-1] += (" " + line)  
    return chapter_list

In [133]:
scarlet_part_1 = get_chapters("A Study in Scarlet".upper(), "PART I", scarlet[54:2585])
scarlet_part_2 = get_chapters("A Study in Scarlet".upper(), "PART II", scarlet[2601:])
scarlet_part = scarlet_part_1 + scarlet_part_2
records += scarlet_part

CHAPTER I Mr. Sherlock Holmes
CHAPTER II The Science Of Deduction
CHAPTER III The Lauriston Garden Mystery
CHAPTER IV What John Rance Had To Tell
CHAPTER V Our Advertisement Brings A Visitor
CHAPTER VI Tobias Gregson Shows What He Can Do
CHAPTER VII Light In The Darkness
CHAPTER I On The Great Alkali Plain
CHAPTER II The Flower Of Utah
CHAPTER III John Ferrier Talks With The Prophet
CHAPTER IV A Flight For Life
CHAPTER V The Avenging Angels
CHAPTER VI A Continuation Of The Reminiscences Of John Watson, M.D.
CHAPTER VII The Conclusion


In [134]:
# The Sign of the Four
four = text[4961:9638]
with open('A Sign of the Four'.upper() + '.txt', "w") as file:
    file.write("\n".join(four))
four_part = get_chapters("A Sign of the Four".upper(), "NO PART", four[34:])
records += four_part

CHAPTER I The Science of Deduction
CHAPTER II The Statement of the Case
CHAPTER III In Quest of a Solution
CHAPTER IV The Story of the Bald-Headed Man
CHAPTER V The Tragedy of Pondicherry Lodge
CHAPTER VI Sherlock Holmes Gives a Demonstration
CHAPTER VII The Episode of the Barrel
CHAPTER VIII The Baker Street Irregulars
CHAPTER IX A Break in the Chain
CHAPTER X The End of the Islander
CHAPTER XI The Great Agra Treasure
CHAPTER XII The Strange Story of Jonathan Small


In [124]:
def parse_no_part(story_name, page_start, page_end, content_start, content_end, no_chapter=False, chapter_name=None):
    story_name = story_name.replace('"', '%')
    with open(story_name+".txt", "w") as file:
        file.write("\n".join(text[page_start: page_end]))
    return get_chapters(story_name, "NO PART", text[content_start: content_end], no_chapter, chapter_name)

In [136]:
# THE ADVENTURES OF SHERLOCK HOLMES 
bohemia_part = parse_no_part("A SCANDAL IN BOHEMIA", 9649, 11219, 9674,11219)
short_titles = ["The Red-Headed League",
               "A Case of Identity",
               "The Boscombe Valley Mystery",
               "The Five Orange Pips",
               "The Man with the Twisted Lip",
               "The Adventure of the Blue Carbuncle",
               "The Adventure of the Speckled Band",
               "The Adventure of the Engineer's Thumb",
               "The Adventure of the Noble Bachelor",
              "The Adventure of the Beryl Coronet",
               "The Adventure of the Copper Beeches"]
short_titles = [ele.upper() for ele in short_titles]
short_stories_parts = list([])
start_idx = 10778
end_idx = 21785
sen_idx = start_idx
num_titles = len(short_titles)
for i in range(num_titles):
    if i != num_titles - 1:
        while short_titles[i+1] not in text[sen_idx]:
            sen_idx += 1
        short_stories_parts.append(parse_no_part(short_titles[i], start_idx, sen_idx-1, start_idx+1, sen_idx-1, True))
        start_idx = sen_idx
    else:
        short_stories_parts.append(parse_no_part(short_titles[-1], start_idx, end_idx, start_idx+1, end_idx, True))
adventure_parts = [bohemia_part] + short_stories_parts 

CHAPTER I
CHAPTER II
CHAPTER III


In [137]:
for chapter in adventure_parts:
    records += chapter

In [108]:
# MEMORIS 
short_titles = ["Silver Blaze",
               "The Yellow Face",
               "The Stock-Broker's Clerk",
               'The "Gloria Scott"',
               "The Musgrave Ritual",
               "The Reigate Squires",
               "The Crooked Man",
               "The Resident Patient",
               "The Greek Interpreter",
               "The Naval Treaty",
               "The Final Problem"]
short_titles = [ele.upper() for ele in short_titles]
short_stories_parts = list([])
start_idx = 21796
end_idx = 31855
sen_idx = start_idx
num_titles = len(short_titles)
for i in range(num_titles):
    if i != num_titles - 1:
        while short_titles[i+1] not in text[sen_idx]:
            sen_idx += 1
        short_stories_parts.append(parse_no_part(short_titles[i], start_idx, sen_idx-1, start_idx+1, sen_idx-1, True))
        start_idx = sen_idx
    else:
        short_stories_parts.append(parse_no_part(short_titles[-1], start_idx, end_idx, start_idx+1, end_idx, True))
memoirs_parts = short_stories_parts

In [138]:
for chapter in memoirs_parts:
    records += chapter

In [111]:
# RETURN
short_titles = [
               'The Adventure of the Empty House',
               'The Adventure of the Norwood Builder',
               'The Adventure of the Dancing Men',
               'The Adventure of the Solitary Cyclist',
               'The Adventure of the Priory School',
               'The Adventure of Black Peter',
               'The Adventure of Charles Augustus Milverton',
               'The Adventure of the Six Napoleons',
               'The Adventure of the Three Students',
               'The Adventure of the Golden Pince-Nez',
               'The Adventure of the Missing Three-Quarter',
               'The Adventure of the Abbey Grange',
               'The Adventure of the Second Stain']
short_titles = [ele.upper() for ele in short_titles]
short_stories_parts = list([])
start_idx = 31866
end_idx = 44779
sen_idx = start_idx
num_titles = len(short_titles)
for i in range(num_titles):
    if i != num_titles - 1:
        while short_titles[i+1] not in text[sen_idx]:
            sen_idx += 1
        short_stories_parts.append(parse_no_part(short_titles[i], start_idx, sen_idx-1, start_idx+1, sen_idx-1, True))
        start_idx = sen_idx
    else:
        short_stories_parts.append(parse_no_part(short_titles[-1], start_idx, end_idx, start_idx+1, end_idx, True))
return_parts = short_stories_parts

In [139]:
for chapter in return_parts:
    records += chapter

In [140]:
# THE HOUND OF THE BASKERVILLES
hound_part = parse_no_part("THE HOUND OF THE BASKERVILLES", 44785, 51843, 44821, 51843)
records += hound_part

CHAPTER I Mr. Sherlock Holmes
CHAPTER II The Curse of the Baskervilles
CHAPTER III The Problem
CHAPTER IV Sir Henry Baskerville
CHAPTER V Three Broken Threads
CHAPTER VI Baskerville Hall
CHAPTER VII The Stapletons of Merripit House
CHAPTER VIII First Report of Dr. Watson
CHAPTER IX Second Report of Dr. Watson
CHAPTER X Extract from the Diary of Dr. Watson
CHAPTER XI The Man on the Tor
CHAPTER XII Death on the Moor
CHAPTER XIII Fixing the Nets
CHAPTER XIV The Hound of the Baskervilles
CHAPTER XV A Retrospection


In [141]:
# THE VALLEY OF FEAR
with open('THE VALLEY OF FEAR.txt', "w") as file:
    file.write("\n".join(text[51848:]))

# part 1
valley_part_1 = get_chapters("THE VALLEY OF FEAR", "PART I", text[51901:55177])
valley_part_2 = get_chapters("THE VALLEY OF FEAR", 'PART II', text[55193:58847])
valley_part = valley_part_1 + valley_part_2
records += valley_part

CHAPTER I The Warning
CHAPTER II Sherlock Holmes Discourses
CHAPTER III The Tragedy of Birlstone
CHAPTER IV Darkness
CHAPTER V The People Of the Drama
CHAPTER VI A Dawning Light
CHAPTER VII The Solution
CHAPTER I The Man
CHAPTER II The Bodymaster
CHAPTER III Lodge 341, Vermissa
CHAPTER IV The Valley of Fear
CHAPTER V The Darkest Hour
CHAPTER VI Danger
CHAPTER VII The Trapping of Birdy Edwards
CHAPTER VIII Epilogue


In [120]:
# HIS LAST BOW
wisteria_part = parse_no_part('The Adventure of Wisteria Lodge'.upper(), 58882, 60204, 58906, 60204)
cardboard_part = parse_no_part('THE ADVENTURE OF THE CARDBOARD BOX', 60209, 61087, 60210, 61087, True)
circle_part = parse_no_part("THE ADVENTURE OF THE RED CIRCLE", 61092, 61970, 61116, 61970)
short_titles = [
              'The Adventure of the Bruce-Partington Plans',
               'The Adventure of the Dying Detective',
               'The Disappearance of Lady Frances Carfax',
               'The Adventure of the Devil\'s Foot',
               'His Last Bow']
short_titles = [ele.upper() for ele in short_titles]
short_stories_parts = list([])
start_idx = 61976
end_idx = 66746
sen_idx = start_idx
num_titles = len(short_titles)
for i in range(num_titles):
    if i != num_titles - 1:
        while short_titles[i+1] not in text[sen_idx]:
            sen_idx += 1
        short_stories_parts.append(parse_no_part(short_titles[i], start_idx, sen_idx-1, start_idx+1, sen_idx-1, True))
        start_idx = sen_idx
    else:
        short_stories_parts.append(parse_no_part(short_titles[-1], start_idx, end_idx, start_idx+1, end_idx, True))
last_row_parts = [wisteria_part, cardboard_part, circle_part] + short_stories_parts

CHAPTER I The Singular Experience of Mr. John Scott Eccles
CHAPTER II The Tiger of San Pedro
CHAPTER I Part One
CHAPTER II Part Two


In [142]:
for chapter in last_row_parts:
    records += chapter

In [121]:
# CASE_BOOK
short_titles = [
               'The Illustrious Client',
               'The Blanched Soldier',
               'The Adventure Of The Mazarin Stone',
               'The Adventure of the Three Gables',
               'The Adventure of the Sussex Vampire',
               'The Adventure of the Three Garridebs',
               'The Problem of Thor Bridge',
               'The Adventure of the Creeping Man',
               'The Adventure of the Lion\'s Mane',
               'The Adventure of the Veiled Lodger',
               'The Adventure of Shoscombe Old Place',
               'The Adventure of the Retired Colourman']
short_titles = [ele.upper() for ele in short_titles]
short_stories_parts = list([])
start_idx = 66826
end_idx = 76729
sen_idx = start_idx
num_titles = len(short_titles)
for i in range(num_titles):
    if i != num_titles - 1:
        while short_titles[i+1] not in text[sen_idx]:
            sen_idx += 1
        short_stories_parts.append(parse_no_part(short_titles[i], start_idx, sen_idx-1, start_idx+1, sen_idx-1, True))
        start_idx = sen_idx
    else:
        short_stories_parts.append(parse_no_part(short_titles[-1], start_idx, end_idx, start_idx+1, end_idx, True))
last_row_parts = [wisteria_part, cardboard_part, circle_part] + short_stories_parts

In [143]:
for chapter in last_row_parts:
    records += chapter

In [126]:
# PREFACE
pf_bow = parse_no_part("PREFACE OF HIS LAST BOW", 58858, 58877, 58859, 58877, True)
pf_case = parse_no_part("PREFACE OF CASE-BOOK", 66756, 66820, 66785, 66820, True)

In [144]:
records += pf_bow
records += pf_case

In [145]:
len(records)

130